lib/gssapi/krb5/store_cred.c

```diff
@@ -46,125 +112,143 @@ _gsskrb5_store_cred_into(OM_uint32         *minor_status,
-	if (ret) {
-	    if (default_cred)
-		ret = krb5_cc_default(context, &id);
-	    else if (def_type &&
-		     krb5_cc_support_switch(context, def_type)) {
-		ret = krb5_cc_new_unique(context, def_type, NULL, &id);
-		overwrite_cred = 1;
-	    } else
-		ret = 0; /* == GSS_C_NO_CRED */
-	}
+    if (cs_ccache_name) {
+        /*
+         * Not the default ccache.
+         *
+         * Therefore not a collection type cache.
+         *
+         * Therefore there's no question of switching the primary ccache.
+         *
+         * Therefore we reset default_cred.
+         *
+         * XXX Perhaps we should fail in this case if default_cred is true.
+         */
+        default_cred = 0;
+	ret = krb5_cc_resolve(context, cs_ccache_name, &id);
+    } else {
+        const char *cctype = NULL;
+
+        /*
+         * Use the default ccache, and if it's a collection, switch it if
+         * default_cred is true.
+         */
+        ret = krb5_cc_default(context, &id);
+        if (ret == 0) {
+	    cctype = krb5_cc_get_type(context, id);
+            if (krb5_cc_support_switch(context, cctype)) {
+                /* The default ccache is a collection type */
+
+                krb5_cc_close(context, id);
+                id = NULL;
+
+                /* Find a matching ccache or create a new one */
+                ret = ccache_match(context, input_cred->principal,
+                                   cctype, &id);
+                if (ret || id == NULL) {
+                    /* Since the ccache is new, just store unconditionally */
+                    overwrite_cred = 1;
+                    ret = krb5_cc_new_unique(context, cctype, NULL, &id);
+                }
+            }
+        }
     }
 
     if (ret || id == NULL) {
-	HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
+	HEIMDAL_MUTEX_unlock(&input_cred->cred_id_mutex);
 	*minor_status = ret;
 	return ret == 0 ? GSS_S_NO_CRED : GSS_S_FAILURE;
     }
 
+    /*
+     * If the new creds are for a different principal than we had before,
+     * overwrite.
+     */
+    if (!overwrite_cred && !same_princ(context, id, input_cred->ccache))
+        overwrite_cred = 1;
+
     if (!overwrite_cred) {
-        /* If current creds are expired or near it, overwrite */
+        /*
+         * If current creds are for the same princ as we already had creds for,
+         * and the new creds live longer than the old, overwrite.
+         */
         ret = krb5_cc_get_lifetime(context, id, &exp_current);
         if (ret != 0 || exp_new > exp_current)
             overwrite_cred = 1;
     }
 
     if (!overwrite_cred) {
         /* Nothing to do */
+        HEIMDAL_MUTEX_unlock(&input_cred->cred_id_mutex);
         krb5_cc_close(context, id);
-        HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
         *minor_status = 0;
         return GSS_S_DUPLICATE_ELEMENT;
     }
 
-    ret = krb5_cc_initialize(context, id, cred->principal);
+    ret = krb5_cc_initialize(context, id, input_cred->principal);
```

Comment:
Sadly, for the default fcache, this likely truncates the file, leaving concurrent users briefly out of luck. Fixing this, might also make it possible to fix "kinit -R", ...

Reply:
We'll fix this separately.